WordPress/openverse
# Initialize Provider store data, UI store and flags store from cookies in a plugin

## Fixes



Fixes #4230 by @obulat

## Description


_One of the main purposes of this PR is to remove the `ssrRef` from the provider store because it's not available in Nuxt 3.
While removing that, I also wanted to simplify the code we run on each page navigation by moving what's possible to the app start on the server (the `init-stores` plugin)._

This PR adds an `init-stores` plugin that initializes the store state at the app's start, instead of sometimes doing it at every page navigation in `middleware`.

Now, all of the `initFromCookies` calls (for the UI store and the flags store) are done in the `init-stores` plugin.
The flags store's `initFromQuery` is still called from the `middleware` in case the query changes between requests.

This plugin also initializes the providers' data. ~~It also keeps track of the date when the providers were updated and if more than 60 minutes have passed since the last update, refetches the provider data. This is the way we implemented this before, but I'm not sure if it's necessary: I don't think we have users that stay on the site for longer than 60 minutes without refreshing the page. Once you refresh the page, the `init-stores` plugin runs and you get the up-to-date provider data anyways. @zackkrida, do you think we should still keep the update-if-60-minutes-passed functionality for provider data?~~ The providers data is initialized once per request, instead of updating it every 60 minutes.

Some of the data that was saved in `env` property in `nuxt.config` was moved to `publicRuntimeConfig` as `env` is not used in Nuxt 3.

## Update
This PR also adds the `sourceName` and `providerName` properties to the media item when the item is fetched (for single item and the search results) instead of when it's requested. This allows to simplify code using the media: we no longer need to call the provider store in component or `_addProviderName` in the single result store.

To fix the Playwright tests in this PR, I also had to update the stats tape, split the cookies that are saved in `preparePageForTests` between the `features` and `sessionFeatures` (e.g. `fetch_sensitive`) cookies, and add waits to some of the tests that were taking the snapshots before the page rendered.

## Testing Instructions

The app should work as expected; there should be no changes. There will be no calls to `initFromCookies` (ui store and flags store), but you would have to add a log in the app to see that.

The CI should pass, too.

## Checklist



[best_practices]:
  https://git-scm.com/book/en/v2/Distributed-Git-Contributing-to-a-Project#_commit_guidelines

## Developer Certificate of Origin


<details>
<summary>Developer Certificate of Origin</summary>

```
Developer Certificate of Origin
Version 1.1

Copyright (C) 2004, 2006 The Linux Foundation and its contributors.
1 Letterman Drive
Suite D4700
San Francisco, CA, 94129

Everyone is permitted to copy and distribute verbatim copies of this
license document, but changing it is not allowed.


Developer's Certificate of Origin 1.1

By making a contribution to this project, I certify that:

(a) The contribution was created in whole or in part by me and I
    have the right to submit it under the open source license
    indicated in the file; or

(b) The contribution is based upon previous work that, to the best
    of my knowledge, is covered under an appropriate open source
    license and I have the right under that license to submit that
    work with modifications, whether created in whole or in part
    by me, under the same open source license (unless I am
    permitted to submit under a different license), as indicated
    in the file; or

(c) The contribution was provided directly to me by some other
    person who certified (a), (b) or (c) and I have not modified
    it.

(d) I understand and agree that this project and the contribution
    are public and that a record of the contribution (including all
    personal information I submit with it, including my sign-off) is
    maintained indefinitely and may be redistributed consistent with
    this project or the open source license(s) involved.
```

</details>
